Fix for eager loading with absent records

al2o3cr opened the pull request:
Fixes the issue discussed below where preloading with a Hash fails if one of the intermediate objects doesn't exist.

If `records` passed to is empty or `[nil]`, skip deciding whether the downstream associations are polymorphic or not.

I'm not sure if it's possible for `records` to have a leading `nil` or not - something like `[nil, <SomeRecord#123>]`. If it is, `find` would be better than `first`.

---
The test added here currently fails with:
```
CascadedEagerLoadingTest#test_eager_association_loading_with_missing_first_record:
NoMethodError: undefined method `_reflect_on_association' for NilClass:Class
    activerecord/lib/active_record/associations/preloader.rb:122:in `block in preloaders_for_hash'
    activerecord/lib/active_record/associations/preloader.rb:115:in `each'
    activerecord/lib/active_record/associations/preloader.rb:115:in `flat_map'
    activerecord/lib/active_record/associations/preloader.rb:115:in `preloaders_for_hash'
    activerecord/lib/active_record/associations/preloader.rb:104:in `preloaders_on'
    activerecord/lib/active_record/associations/preloader.rb:126:in `block (2 levels) in preloaders_for_hash'
    activerecord/lib/active_record/associations/preloader.rb:125:in `each'
    activerecord/lib/active_record/associations/preloader.rb:125:in `flat_map'
    activerecord/lib/active_record/associations/preloader.rb:125:in `block in preloaders_for_hash'
    activerecord/lib/active_record/associations/preloader.rb:115:in `each'
    activerecord/lib/active_record/associations/preloader.rb:115:in `flat_map'
    activerecord/lib/active_record/associations/preloader.rb:115:in `preloaders_for_hash'
    activerecord/lib/active_record/associations/preloader.rb:104:in `preloaders_on'
    activerecord/lib/active_record/associations/preloader.rb:93:in `block in preload'
    activerecord/lib/active_record/associations/preloader.rb:92:in `each'
    activerecord/lib/active_record/associations/preloader.rb:92:in `flat_map'
    activerecord/lib/active_record/associations/preloader.rb:92:in `preload'
    activerecord/lib/active_record/relation.rb:627:in `block in preload_associations'
    activerecord/lib/active_record/relation.rb:625:in `each'
    activerecord/lib/active_record/relation.rb:625:in `preload_associations'
    activerecord/lib/active_record/relation.rb:661:in `block in exec_queries'
    activerecord/lib/active_record/relation.rb:676:in `skip_query_cache_if_necessary'
    activerecord/lib/active_record/relation.rb:645:in `exec_queries'
    activerecord/lib/active_record/relation.rb:508:in `load'
    activerecord/lib/active_record/relation.rb:238:in `records'
    activerecord/lib/active_record/relation.rb:233:in `to_ary'
    activerecord/test/cases/associations/cascaded_eager_loading_test.rb:165:in `test_eager_association_loading_with_missing_first_record'
```

The cause appears to be code added in https://github.com/rails/rails/commit/75ef18c67c29b1b51314b6c8a963cee53394080b which assumes that `preloaders_for_hash` will always be called with `records` having a first value that isn't nil.

The test breaks this assumption, since the `Post` with ID 3 has no author. `records` is passed in as `[nil]`.

I haven't located an example in the existing fixtures, but it's certainly possible to imagine a scenario where `records` could have a *leading* `nil` followed by ActiveRecord objects.

/cc @tenderlove @eileencodes since this is breaking a bunch of our tests :)

al2o3cr committed:
WIP: failing test for eager loading

al2o3cr committed:
only reflect on non-nil objects

eileencodes commented:
@al2o3cr is this ready? Looks like it fixed the problem in our app. If it's good I can merge this.

al2o3cr commented:
@eileencodes I'd like to see somebody who understands more about how preload is working (particularly with the new polymorphic handling) before this gets merged; in particular, is there any case where `records.first` is `nil` but more loading should be done?

(in the meantime there's https://github.com/github/github/pull/97755 to get CI closer to green)

eileencodes commented:
Hey @robotdana you did the original implementation. Thoughts on this PR / bug fix?

Also cc/ @tenderlove @rafaelfranca

robotdana commented:
I'm a little unclear about why `nil` is in the list of records, but we're definitely not caring about specifically first so much as 'the class of what we're preloading into'

https://github.com/rails/rails/blob/c38acb96d3706def3a8081ea611e43715ed5f5d8/activerecord/lib/active_record/associations/preloader.rb#L158 does seem to imply that the list can contain nils in arbitrary locations and we should continue in the case of nils, rather than just an empty array being the issue, so checking `first_record = records.find(&:itself) instead of `records.first` might be safer. (can we use `#itself`? what ruby version did that come in?)

robotdana commented on activerecord/lib/active_record/associations/preloader.rb:
```diff
@@ -117,12 +117,15 @@ def preloaders_for_hash(association, records, scope, polymorphic_parent)
+              reflection = first_record.class._reflect_on_association(parent)
+              polymorphic_parent = reflection && reflection.options[:polymorphic]
+
+              loaders.concat Array.wrap(child).flat_map { |assoc|
```
I think this should be outside the if statement? at least it wasn't part of the change I added

al2o3cr replied:
@robotdana the block passed to `flat_map` references `polymorphic_parent`

al2o3cr committed:
Fix assertion order

al2o3cr committed:
Only recurse when there's more preloading to do

`preloaders_for_one` drops `nil`s in the incoming array, in
`grouped_records`. As a result, if it returns a loader with a non-empty
`preloaded_records` then `records` must have at least one non-nil
element.

`preloaders_on` returns an empty array if it's passed an empty array and
does not modify any other state, so it only needs to be called if
preloaded records are present.

Taken together: it's only necessary to calculate `polymorphic_parent` if
`recs` is non-empty, and then is always possible to do so since `records`
must have at least one non-`nil` element.

al2o3cr commented:
I rewrote the implementation to be more careful around `nil`s. From the commit:

`preloaders_for_one` drops `nil`s in the incoming array, in `grouped_records`. As a result, if it returns a loader with a non-empty `preloaded_records` then `records` must have at least one non-`nil` element.

`preloaders_on` returns an empty array if it's passed an empty array and does not modify any other state, so it only needs to be called if preloaded records are present.

Taken together: it's only necessary to calculate `polymorphic_parent` if `recs` is non-empty, and then is always possible to do so since `records` must have at least one non-`nil` element.

al2o3cr commented:
@eileencodes This is ready for review; the CI failure looks like some kind of chromedriver hiccup.

matthewd commented:
FYI, this feels like it might have overlap with #33896, and possibly #32136.

al2o3cr commented:
@matthewd It's adjacent but not exactly related; we ran into this bug in tests of code that does things very similar to #32136 to hydrate associations for API responses.

Both of those PRs touch code that's downstream from `grouped_records` (via `preloader_for` [here](https://github.com/rails/rails/blob/2df65b6fae6d5356ba783809a3c9a598f1539e3e/activerecord/lib/active_record/associations/preloader.rb#L144)) so they won't ever see the emptyish arrays this PR handles.

rafaelfranca commented on activerecord/test/cases/associations/cascaded_eager_loading_test.rb:
```diff
@@ -161,6 +161,11 @@ def test_eager_association_loading_where_first_level_returns_nil
   end
 
+  def test_eager_association_loading_with_missing_first_record
+    posts = Post.where(id: 3).preload(author: { comments: :post }).to_a
```
It is not clear to me how this test is exercising this problem. Is the problem that The post with id 3 doesn't exist of if that the author that is missing, or the comment?

al2o3cr replied:
@rafaelfranca The post with ID 3 has `author_id` nil, so `preloaders_for_hash` gets called with an empty `records` and `association` set to `{comments: :post}`

eileencodes commented:
Any objections to this @matthewd @rafaelfranca? I'd like to get this merged so we can get our Rails 6 build greener (this is currently the majority of failures).

faucct commented:
Please look into an alternative solution to this - https://github.com/rails/rails/pull/33938
